packages/recommend-react/src/useRelatedProducts.ts

```diff
@@ -18,34 +19,47 @@ export function useRelatedProducts<TObject>({
   recommendClient,
   threshold,
   transformItems: userTransformItems,
-}: GetRelatedProductsProps<TObject>) {
-  const [result, setResult] = useState<GetRecommendationsResult<TObject>>({
-    recommendations: [],
-  });
+  initialState: userInitialState,
+}: GetRelatedProductsProps<TObject> & {
+  initialState?: InitialResults<TObject>;
+}) {
+  const isFirstRenderRef = useRef(false);
+
   const { status, setStatus } = useStatus('loading');
   const objectIDs = useStableValue(userObjectIDs);
   const transformItems = useStableValue(userTransformItems);
+  const initialState = useStableValue(userInitialState);
   const queryParameters = useStableValue(userQueryParameters);
   const fallbackParameters = useStableValue(userFallbackParameters);
 
+  const initialRecommendationsResult = initialState ?? { recommendations: [] };
+  const [recommendationsResult, setRecommendationsResult] = useState<
+    GetRecommendationsResult<TObject>
+  >(initialRecommendationsResult);
+
   useAlgoliaAgent({ recommendClient });
 
   useEffect(() => {
-    setStatus('loading');
-    getRelatedProducts({
-      fallbackParameters,
-      indexName,
-      maxRecommendations,
-      objectIDs,
-      queryParameters,
-      recommendClient,
-      threshold,
-      transformItems,
-    }).then((response) => {
-      setResult(response);
-      setStatus('idle');
-    });
+    if (!initialState || isFirstRenderRef.current) {
+      setStatus('loading');
+      getRelatedProducts({
+        fallbackParameters,
+        indexName,
+        maxRecommendations,
+        objectIDs,
+        queryParameters,
+        recommendClient,
+        threshold,
+        transformItems,
+      }).then((response) => {
+        setRecommendationsResult(response);
+        setStatus('idle');
+      });
+    } else {
+      isFirstRenderRef.current = true;
+    }
```

Comment:
I _think_ there are some broken behaviors in the way that the ref is reset. It needs to be reset either way after this effect.

I also think that having an intermediate `shouldFetch` variable is helpful here to make the intent clear why we have this guard.

```suggestion
    const shouldFetch = !initialState || !isFirstRenderRef.current;

    if (shouldFetch) {
      setStatus('loading');
      getRecommendations({
        fallbackParameters,
        indexName,
        maxRecommendations,
        model,
        objectIDs,
        queryParameters,
        recommendClient,
        threshold,
        transformItems,
      }).then((response) => {
        setResult(response);
        setStatus('idle');
      });
    }

    isFirstRenderRef.current = false;
```

Reply:
✅